### PR TITLE
feat: support group slug on relation creation

### LINKED ISSUE #13: Support group slug in resources definition
**Is your feature request related to a problem? Please describe.**
So far we only support ID for the resources definition. Our hook will create a new relation in the DB with UUID. There is a need to support Group slug as well in the resource creation.

**Describe the solution you'd like**
If Group slug is passed, we could transform the group slug to uuid before creating a new relation.

### PR DESCRIPTION
**Pull request by mabdh**
I found that the possible way to solve this is to convert group slug to group ID at this point
https://github.com/goto/shield/blob/main/core/relation/service.go#L44

But adding the case here introduce import cycle. This is because the relation package is being used by all domains package.

godepgraph

Solution:

1. Move the transformation of slug to id and user e-mail to user id outside relation package
2. Refactor the code. Inverse the dependency, so domains won't depend on relation package but the other way around, relation is a higher packages that has dependency to all domains

Decision:
**Solution 1**
- It is simpler to implement compare to Solution 2
---
- I added an adapter for all transformation outside the core package level
- Decouple `relationService.Create` and relation transformation. All relations that goes to `relationService.Create` params should already be conditioned (only accept uuid) and the transformation is being done before calling `relationService.Create`
- Fixing end to end tests and added a new test case where using group slug in the hook
- Use `.mockery.yaml` for generating mocks

### CODE REVIEW

File: internal/adapter/relation.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		relationService: relationService,
+	}
+}
+
+func (a Relation) TransformRelation(ctx context.Context, rlt relation.RelationV2) (relation.RelationV2, error) {
+	rel := rlt
+
+	// If Principal is a user, then we will get ID for that user as Subject.ID
+	if rel.Subject.Namespace == schema.UserPrincipal || rel.Subject.Namespace == "user" {
+		fetchedUser, err := a.userService.GetByEmail(ctx, rel.Subject.ID)
```
Comment (ishanarya0) [inline internal/adapter/relation.go]: could there be a use case where we have to allow user uuid and email interchangebly? @AkarshSatija @mabdh 

if yes, let's add it right away?

File: internal/adapter/relation.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+			groupID = fetchedGroup.ID
+		}
+		rel.Object.NamespaceID = schema.GroupNamespace
+		rel.Object.ID = groupID
+	}
+
+	return rel, nil
+}
+
+// // If Principal is a user, then we will get ID for that user as Subject.ID
```
Comment (ishanarya0) [inline internal/adapter/relation.go]: are these. comments needed?

**Review by ishanarya0 — CHANGES_REQUESTED** [review]

File: internal/adapter/relation.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		relationService: relationService,
+	}
+}
+
+func (a Relation) TransformRelation(ctx context.Context, rlt relation.RelationV2) (relation.RelationV2, error) {
+	rel := rlt
+
+	// If Principal is a user, then we will get ID for that user as Subject.ID
+	if rel.Subject.Namespace == schema.UserPrincipal || rel.Subject.Namespace == "user" {
+		var userID = rel.Subject.ID
```
Comment (ishanarya0) [inline internal/adapter/relation.go:37]: Can we add a test where shield should allow relation creation when subject namespace is user and subject id is a user uuid?

**Review by ishanarya0 — CHANGES_REQUESTED** [review]

File: internal/adapter/relation.go (hunk 3 above)
Comment (mabdh) [reply internal/adapter/relation.go:37]: yeah we can, let me add that

**Review by ishanarya0 — APPROVED** [review]

**Pull request merged**